shadowsocks/config.go

```diff
@@ -1,76 +1,198 @@
+
+// rolling index give out the index which server will return on next rolling get server
+var roundRobinIndex int
+
+// Config implement the ss config
 type Config struct {
-	Server     interface{} `json:"server"`
-	ServerPort int         `json:"server_port"`
-	LocalPort  int         `json:"local_port"`
-	Password   string      `json:"password"`
-	Method     string      `json:"method"` // encryption method
-	Auth       bool        `json:"auth"`   // one time auth
+	Server          string `json:"server_addr"`     // shadowsocks remote Server address, for multi server split them with comma
+	ServerPort      string `json:"server_port"`     // shadowsocks remote Server port, split with comma when multi user is enabled
+	Local           string `json:"local_addr"`      // shadowsocks local socks5 Server address
+	LocalPort       int    `json:"local_port"`      // shadowsocks local socks5 Server port
+	Password        string `json:"password"`        // shadowsocks remote server password, for multi server password should plase in order and split eith comma
+	Method          string `json:"method"`          // shadowsocks encryption method, split by comma if multi server is enabled
+	Timeout         int    `json:"timeout"`         // shadowsocks connection timeout
+	MultiServerMode string `json:"MultiServerMode"` // shadowsocks client multi-server access mode: fastest,round-robin,dissable
+	DNSServer       string `json:"dns_server"`      // shadowsocks remote Server DNS server option, if set to nil, the system DNS will be uesd for domain lookup
+
+	servers []string
+	methods []string // methods array, match the server port
 
 	// following options are only used by server
-	PortPassword map[string]string `json:"port_password"`
-	Timeout      int               `json:"timeout"`
+	PortPassword map[string]string `json:"port_password"` // shadowsocks mutli user password
 
 	// following options are only used by client
+	ServerPassword map[string]string `json:"server_password"` // shadowsocks local mutli server password
+
+	// TODO unsupported functions
+	//"fast_open":false,
+	//"tunnel_remote":"8.8.8.8",
+	//"tunnel_remote_port":53,
+	//"tunnel_port":53,
+}
+
+// String return the ss config content in string
+func (c *Config) String() string {
+	return fmt.Sprintf("Server: %s, ServerPort: %s, Local: %s, LocalPort: %d, Password: %s, Method: %s, Timeout: %d, portpwds: %v, serverpwds: %v, multi-server module:%v",
+		c.Server, c.ServerPort, c.Local, c.LocalPort, c.Password, c.Method, c.Timeout, c.PortPassword, c.ServerPassword, c.MultiServerMode)
+}
+
+// ConfOption define the config options
+type ConfOption func(c *Config)
+
+// NewConfig use the option to generate the ss config
+func NewConfig(opts ...ConfOption) *Config {
+	var config = Config{
+		PortPassword:   make(map[string]string),
+		ServerPassword: make(map[string]string),
+		servers:        make([]string, 1),
+	}
+	for _, v := range opts {
+		v(&config)
+	}
+
+	servers := config.GetServerArray()
```

Comment:
following 5 lines have been reconstruct into function-> makeupServers

Reply:
@dszhengyu fixed in 8796767